s2a: Add gRPC S2A

rmehta19 opened the pull request:
Add S2A Java client to gRPC Java.

Context: https://github.com/google/s2a-go/blob/main/README.md

rmehta19 commented:
@matthewstevenson88, I made 2 changes and it looks like 2/3 linux runs are passing. tests(11) is failing with an [error](https://github.com/grpc/grpc-java/actions/runs/8788883897/job/24117364595?pr=11113#step:7:1753) in code that was not affected by this PR, so I don't think that failure is related. 

The changes:
- The build originally failed because the generated grpc code in this PR did not match the code generated by the codegen plugin. I couldn't easily compile the codegen myself(due to reasons explained in: https://github.com/grpc/grpc-java/issues/7690). So instead I used the latest published binary available on https://mvnrepository.com/artifact/io.grpc/protoc-gen-grpc-java (1.63) to generate S2AServiceGrpc.java, then I changed line 8 & 9 of S2AServiceGrpc.java to match other generated code in this repo (ex: https://github.com/grpc/grpc-java/blob/fb9a10809f5106333f447fe0cda1c761d5846f72/alts/src/generated/main/grpc/io/grpc/alts/internal/HandshakerServiceGrpc.java#L8-L9). This resolved the errors thrown by https://github.com/grpc/grpc-java/blob/3c867c9ab7db99509830d3a722fb6f9098a1217e/.github/workflows/testing.yml#L65-L66

- After resolving above, for some reason when IntegrationTest.java is in package io.grpc.s2a, GetAuthenticationMechanismsTest.java fails because the flag is not being set by JCommander. When I moved IntegrationTest.java into the same package as  GetAuthenticationMechanismsTest.java: io.grpc.s2a.handshaker, this error does not happen.

matthewstevenson88 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,132 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Configures gRPC to use S2Av2 for transport security when establishing a secure channel. Only for
```
nit: Since this is the top-level API, can we say `S2A` rather than `S2Av2` here?

rmehta19 replied:
Done.

matthewstevenson88 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2ATrustManager.java:
```diff
@@ -0,0 +1,152 @@
+import javax.net.ssl.X509TrustManager;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/** Handles requests on verification of peer's certificates. */
```
nit: This sentence is a bit confusing, could we maybe say "Offloads verification of the peer certificate chain to S2A."?

rmehta19 replied:
Done.

matthewstevenson88 replied:
Thanks!

matthewstevenson88 commented on s2a/src/main/proto/grpc/gcp/common.proto:
```diff
@@ -0,0 +1,79 @@
+// Copyright 2022 Google LLC
```
Just to confirm, this proto will be removed from this PR once the other PR is merged?

matthewstevenson88 replied:
Makes sense, thanks for clarifying!

rmehta19 replied:
Done

matthewstevenson88 commented:
LGTM, thanks @rmehta19! When you have a chance, please remove "draft" to the PR title and can we say a bit more in the CL description including pointing to the S2A-Go README (similar to what we did in the proto PR)?

After that, can we get sign-off from @erm-g and @ejona86?

rmehta19 commented:
> LGTM, thanks @rmehta19! When you have a chance, please remove "draft" to the PR title and can we say a bit more in the CL description including pointing to the S2A-Go README (similar to what we did in the proto PR)?
> 
> After that, can we get sign-off from @erm-g and @ejona86?

Done -- thank you for the review @matthewstevenson88!

erm-g commented on s2a/src/main/java/io/grpc/s2a/handshaker/tokenmanager/SingleTokenFetcher.java:
```diff
@@ -0,0 +1,64 @@
+        description = "The access token used to authenticate to S2A.")
+    private static String accessToken = System.getenv(ENVIRONMENT_VARIABLE);
+
+    public synchronized void reset() {
```
I'm not sure I understand why 'synchronized' is needed here.

rmehta19 replied:
This code doesn't exist anymore, as we removed the dependency on Jcommander and replaced  it with a private field & setter

erm-g commented on s2a/src/test/resources/config.cnf:
```diff
@@ -0,0 +1,17 @@
+distinguished_name = req_distinguished_name
+req_extensions = req_ext
+
+[req_distinguished_name]
```
Could you add the values you entered while generating certs?

rmehta19 replied:
Done, as discussed offline, I updated the README with the the values used to populate these fields. 

@ejona86, does this sound ok to you? @erm-g pointed me to something similar he has done in go: https://github.com/grpc/grpc-go/pull/6670/files#diff-ac85255b49c8d65b03b560d9959b99c6d9338de1400da6fc2d4133a434a376ad

ejona86 commented on s2a/src/test/java/io/grpc/s2a/handshaker/GetAuthenticationMechanismsTest.java:
```diff
@@ -0,0 +1,64 @@
+  @BeforeClass
+  public static void setUpClass() {
+    // Set the token that the client will use to authenticate to the S2A.
+    JCommander.newBuilder().addObject(FLAGS).build().parse(SET_TOKEN);
```
You're using jcommander to just set a static field. This is many more steps than just having a package-private field/setter. Let's not add a dependency on jcommander.

ejona86 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,132 @@
+ * Configures gRPC to use S2A for transport security when establishing a secure channel. Only for
+ * use on the client side of a gRPC connection.
+ */
+@NotThreadSafe
```
Credentials must be thread-safe. Should this be on the Builder instead?

rmehta19 replied:
Yes, the Builder is the one that's not thread safe -- done.

ejona86 reviewed (commented):
To make sure I understand, s2a is the public API, and s2a/channel and s2a/handshaker are internal APIs. Are those internal APIs to be used anywhere, even within Google?

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AGrpcChannelPool.java:
```diff
@@ -0,0 +1,112 @@
+ */
+@ThreadSafe
+public final class S2AGrpcChannelPool implements S2AChannelPool {
+  private static final int MAX_NUMBER_USERS_OF_CACHED_CHANNEL = 100000;
```
hardcoding 100000 seems fine to me, @matthewstevenson88 WDYT?

matthewstevenson88 replied:
If it's OK with you all, I'd prefer to keep it hardcoded until we have a use case that requires us to expose it.

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AHandshakerServiceChannel.java:
```diff
@@ -0,0 +1,195 @@
+public final class S2AHandshakerServiceChannel {
+  private static final ConcurrentMap<String, Resource<Channel>> SHARED_RESOURCE_CHANNELS =
+      Maps.newConcurrentMap();
+  private static final Duration DELEGATE_TERMINATION_TIMEOUT = Duration.ofSeconds(2);
```
Hardcoding these two timeouts is ok with me, keeping it as is might help readability. @matthewstevenson88 WDYT?

matthewstevenson88 replied:
If it's OK with you all, I'd prefer to keep it hardcoded until we have a use case that requires us to expose it.

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AGrpcChannelPool.java:
```diff
@@ -0,0 +1,112 @@
+  public synchronized Channel getChannel() {
+    checkState(state.equals(State.OPEN), "Channel pool is not open.");
+    checkState(
+        numberOfUsersOfCachedChannel >= 0,
```
You are right, thanks for pointing this out -- removed this unnecessary state check.

rmehta19 commented:
> To make sure I understand, s2a is the public API, and s2a/channel and s2a/handshaker are internal APIs. Are those internal APIs to be used anywhere, even within Google?

Correct s2a is the public api, s2a/handshaker and s2a/channel are internal APIs to only be used within the s2a package. 

cc: @matthewstevenson88

erm-g commented on s2a/src/test/java/io/grpc/s2a/handshaker/S2ATrustManagerTest.java:
```diff
@@ -0,0 +1,262 @@
+  private S2AStub stub;
+  private FakeWriter writer;
+  private static final String FAKE_HOSTNAME = "Fake-Hostname";
+  private static final String CLIENT_CERT_PEM =
```
Does this cert have any expiration which can make the test fail in the future?

rmehta19 replied:
Discussed offline -- no changes to be made.

erm-g commented on s2a/src/test/java/io/grpc/s2a/handshaker/FakeS2AServer.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.logging.Logger;
+
+/** A fake S2Av2 server that should be used for testing only. */
+public final class FakeS2AServer extends S2AServiceGrpc.S2AServiceImplBase {
```
For Fake classes - I saw that gRPC java uses mockito quite extensively. Suggesting checking with Eric if this 'fake' style is ok or needs to be migrated to mocks

larry-safran replied:
Fakes are better than mocks, they are just more work.  Eric is driven crazy by how much mocking is currently done, don't use it as a best practice.

erm-g commented on s2a/src/test/java/io/grpc/s2a/handshaker/IntegrationTest.java:
```diff
@@ -0,0 +1,322 @@
+public final class IntegrationTest {
+  private static final Logger logger = Logger.getLogger(FakeS2AServer.class.getName());
+
+  private static final String CERT_CHAIN =
```
Same q about certs - do they have any expiration date?

rmehta19 replied:
Discussed offline -- no changes to be made.

erm-g commented on s2a/src/test/java/io/grpc/s2a/handshaker/IntegrationTest.java:
```diff
@@ -0,0 +1,322 @@
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class IntegrationTest {
```
I believe it should have some prefix like 'S2A'

rmehta19 replied:
Discussed offline -- no changes to be made.

rmehta19 commented on s2a/src/test/java/io/grpc/s2a/handshaker/tokenmanager/SingleTokenAccessTokenManagerTest.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  @Test
+  public void create_noEnvironmentVariable() throws Exception {
```
Discussed offline -- no changes to be made.

rmehta19 commented:
@ejona86, PTAL when you get a chance, and leave any additional feedback. Thank you in advance!

@erm-g & @matthewstevenson88 have approved these changes.

larry-safran commented:
On July 16th @ejona86 replied to an email requesting some major reworking in your implementation.  Has there been any progress on that?  Specifically:

1. No blocking on the event loop
2. Use Async, not the blocking stub
3. Change to a more standard way of using error classes
4. Have your tests specify reasonable deadlines

I've been waiting on reviewing until this rework was done.

From the email:
```
I'm not comfortable with that handlerAdded(). InterruptedException causes loud sirens; there must be no blocking on the event loop. We were promised ALTS would stop blocking, but that was never implemented and it caused my team lots of time on debugging and workarounds. Given you're able to use SSLHandler, I don't think there'd be any strong reason to allow blocking, even initially. I see the blocking stub, and that will probably need to be changed. IllegalArgumentException is just the wrong exception type; that means "the programmer made an error" and it should basically never be caught. But it also means an argument was wrong, and a KeyStoreException seems unrelated to the arguments. Mixing "ChannelHandler next" and "negotiator.newHandler()" also seems suspicious, as it seems you are delegating twice.
```

rmehta19 commented:
Thanks @larry-safran, the item's have been addressed

larry-safran commented on s2a/src/test/java/io/grpc/s2a/handshaker/IntegrationTest.java:
```diff
@@ -0,0 +1,322 @@
+
+  @BeforeClass
+  public static void setUpClass() {
+    System.setProperty("GRPC_EXPERIMENTAL_ENABLE_NEW_PICK_FIRST", "false");
```
Why are you overriding the default value.  Everything should work the same regardless of which pick first is used.

larry-safran replied:
Adding `awaitTermination(10, TimeUnit.SECONDS)` to teardown will get rid of that TSan data race so you won't need this.

rmehta19 replied:
Thanks for the investigation Larry! Changes in https://github.com/grpc/grpc-java/pull/11113/commits/752627a8e5b383f9bb06c5296bd1691a5987f1da.

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AChannelPool.java:
```diff
@@ -0,0 +1,43 @@
+  /**
+   * Retrieves an open channel to the S2A from the channel pool.
+   *
+   * <p>If no channel is available, blocks until a channel can be retrieved from the channel pool.
```
Thanks for pointing that out, updated the comment in https://github.com/grpc/grpc-java/pull/11113/commits/38b0a3a2c254785f375d4c4e9407f6405bf7df08 to reflect the implementation.

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/handshaker/GetAuthenticationMechanisms.java:
```diff
@@ -0,0 +1,60 @@
+   * @return an {@link AuthenticationMechanism} for the given local identity.
+   */
+  static Optional<AuthenticationMechanism> getAuthMechanism(Optional<S2AIdentity> localIdentity) {
+    Optional<AuthenticationMechanism> authMechanism = Optional.empty();
```
Removed this local var in https://github.com/grpc/grpc-java/pull/11113/commits/12586b11508fd5fca7058e8e414428b93d7e45f2

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/handshaker/ProtoUtil.java:
```diff
@@ -0,0 +1,96 @@
+      } catch (IllegalArgumentException e) {
+        continue;
+      }
+      if (versionNumber < minTlsVersion.getNumber() || versionNumber > maxTlsVersion.getNumber()) {
```
Done in https://github.com/grpc/grpc-java/pull/11113/commits/08f83421b3ece4a7e44a84637971451f141f5f9e

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,132 @@
+    @CanIgnoreReturnValue
+    public Builder setLocalSpiffeId(String localSpiffeId) {
+      checkNotNull(localSpiffeId);
+      localIdentity = S2AIdentity.fromSpiffeId(localSpiffeId);
```
Thanks for pointing this out, done in https://github.com/grpc/grpc-java/pull/11113/commits/7f267121aa662379a1c137de27bf0f73c8b76cda

larry-safran commented on s2a/src/test/java/io/grpc/s2a/handshaker/IntegrationTest.java:
```diff
@@ -0,0 +1,322 @@
+      throws InterruptedException {
+    try {
+      SimpleServiceGrpc.SimpleServiceBlockingStub stub =
+          SimpleServiceGrpc.newBlockingStub(channel).withWaitForReady();
```
You should remove the waitForReady.  (As an aside, anytime you use waitForReady you need to add a deadline)

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11113/commits/0e059e4d1a3a4477abe53f7d9fd9f411a9395990

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java:
```diff
@@ -0,0 +1,245 @@
+    static S2AProtocolNegotiator createForClient(
+        S2AChannelPool channelPool, Optional<S2AIdentity> localIdentity) {
+      checkNotNull(channelPool, "Channel pool should not be null.");
+      checkNotNull(localIdentity, "Local identity should not be null on the client side.");
```
Correct, however I think the reason we Optional<S2AIdentity> is because the localIdentity can be empty, in this case S2A will choose a default localIdentity, see https://github.com/grpc/grpc-java/pull/11113/files#diff-7502f77f0608fe8242cff3b41e9c4176bc362d2ca191a0b4a9121b7e4f9496f6R104-R106, for example.

cc: @matthewstevenson88

larry-safran replied:
My point was that you should either reject Optional.empty values like you are null values, or if it is okay to have empty values, then just convert nulls to Optional.empty instead of throwing an IllegalArgumentException.  There are a number of places like this where you do a check for null on an Optional.

rmehta19 replied:
Thanks for explaining, this makes sense. It is ok to have Optional.empty, so I have pushed down the creation of the actual Optional.empty from a null to S2AProtocolNegotiator. Changes in: https://github.com/grpc/grpc-java/pull/11113/commits/b021a2128ec2dc92fa403c190d4aeaa60c259a78

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java:
```diff
@@ -0,0 +1,245 @@
+    public ChannelHandler newHandler(GrpcHttp2ConnectionHandler grpcHandler) {
+      checkNotNull(grpcHandler, "grpcHandler should not be null.");
+      String hostname = getHostNameFromAuthority(grpcHandler.getAuthority());
+      checkNotNull(hostname, "hostname should not be null.");
```
Done in https://github.com/grpc/grpc-java/pull/11113/commits/19583a7b136989d6c1d4b88be814cbe2e3e15cc4

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
```
Done in https://github.com/grpc/grpc-java/pull/11113/commits/3198eec7ca420d098d0c37d793774bf84e72d2d0

larry-safran replied:
You missed doing the same for channelReadComplete

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11113/commits/a8cacb0f14d1235adb47be598b8b49b5096ba778

rmehta19 commented:
Thanks for the review @larry-safran ! Please let me know if there is anything else to address.

larry-safran commented on s2a/src/test/java/io/grpc/s2a/handshaker/IntegrationTest.java:
```diff
@@ -0,0 +1,322 @@
+    server.shutdown();
+    s2aServer.shutdown();
+    s2aDelayServer.shutdown();
+    mtlsS2AServer.shutdown();
```
Change this to `mtlsS2AServer.shutdown(). awaitTermination(10, TimeUnit.SECONDS)`

rmehta19 replied:
Thanks Larry! Done in https://github.com/grpc/grpc-java/pull/11113/commits/752627a8e5b383f9bb06c5296bd1691a5987f1da

ejona86 reviewed (commented):
Some random things I noticed.

ejona86 commented on s2a/src/test/java/io/grpc/s2a/handshaker/IntegrationTest.java:
```diff
@@ -0,0 +1,320 @@
+public final class IntegrationTest {
+  private static final Logger logger = Logger.getLogger(FakeS2AServer.class.getName());
+
+  private static final String CERT_CHAIN =
```
Do they have an expiration date? In all other cases we have documentation on how these were created so we 1) know what they are and 2) know how to recreate them. FakeWriter has even more...

rmehta19 replied:
I have removed the hardcoded certs from this file and FakeWriter. I changed these files to read the certs from the resources folder like the rest of the tests. I regenerated the certs used in these two files and added instructions for how to do so in the resources folder. All of the certs used in the S2A tests expire in 20 yrs. See readme for commands to regenerate.

Done in https://github.com/grpc/grpc-java/pull/11540/commits/19b8f2176abf5f433ccc27d7c79e96260f0e43fb

ejona86 commented on s2a/src/test/resources/client.csr:
```diff
@@ -0,0 +1,16 @@
+-----BEGIN CERTIFICATE REQUEST-----
```
It is weird to commit the csr. That is temporary, right?

rmehta19 replied:
From the [README](https://github.com/grpc/grpc-java/blob/master/s2a/src/test/resources/README.md),the csr is created as part of the steps to regenerate the certs in this folder, which have a lifetime of 20 yrs. This resources directory is similar to the one we have the S2A Go client: https://github.com/google/s2a-go/tree/main/testdata

ejona86 commented on s2a/src/test/resources/config.cnf:
```diff
@@ -0,0 +1,17 @@
+subjectAltName = @alt_names
+
+[alt_names]
+IP.1 = ::
```
What is happening here? How would this ever work?

rmehta19 replied:
This is so that the generated certs have localhost IP addr, so that certificate verification passes in tests

ejona86 replied:
Isn't localhost `::1`?

rmehta19 replied:
This is true. I ran a local test modifying the IP to be `::1` and regenerating these certs, confirming our tests still pass. I'll modify this in a separate PR along with the other comments on the EC cert data: https://github.com/grpc/grpc-java/pull/11540#discussion_r1778823717 .

I'll also update our go client tests for consistency: https://github.com/google/s2a-go/blob/main/testdata/config.cnf

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11583/commits/7a879a2d3c0b97f183a8ec0bae181f7136256bc7

ejona86 replied:
I think that's the wrong conclusion. It means IP matching wasn't actually used. So if it should be and wasn't, then you need to find that hostname matching bug and then we'll also fix this cert. If it was matching already, then the IP wasn't used.

ejona86 replied:
FWIW, we use `managedChannelBuilder.overrideAuthority()` to choose what we expect in the cert in tests. It lets us use one name for address resolution and another for TLS.

rmehta19 replied:
Sorry for the confusion and thanks for pointing out. To be clear, out tests rely on the fact that Common Name is set to localhost (not actually in this config, but when this config is used to create the CSR, it is set. I noted this in the README). So that's why this field was left as ::, and had no effect on tests when set to ::1.

ejona86 replied:
Then let me repeat my initial question: What is happening here? What is it doing?

rmehta19 replied:
We use the config to generate certs to be used for client & S2A when testing mTLS to S2A. This is used in an [integration test](https://github.com/grpc/grpc-java/blob/master/s2a/src/test/java/io/grpc/s2a/internal/handshaker/IntegrationTest.java#L145) (also in S2AChannelCredentialsTest unit test, but that's just a unit test, no channel actually gets established between the client and S2A). As part of mTLS peer cert verification process, it is required that certificate CommonName matches the hostname (which is localhost).

ejona86 replied:
I don't understand. This is alt_names aka SAN, which is not CommonName. And `::` is not `::1` and that would not validate if the client used "localhost". So that all seems irrelevant.

ejona86 replied:
(As I said before, as a hint for integration tests, don't put localhost in the cert, which is actually invalid. Instead, use `managedChannelBuilder.overrideAuthority()`. It is available in all languages; [`WithAuthority()` in Go](https://pkg.go.dev/google.golang.org/grpc#WithAuthority).)

ejona86 commented on s2a/src/test/java/io/grpc/s2a/handshaker/FakeS2AServerTest.java:
```diff
@@ -0,0 +1,265 @@
+  private static final ImmutableList<ByteString> FAKE_CERT_DER_CHAIN =
+      ImmutableList.of(
+          ByteString.copyFrom(
+              new byte[] {'f', 'a', 'k', 'e', '-', 'd', 'e', 'r', '-', 'c', 'h', 'a', 'i', 'n'}));
```
FYI: `"fake-der-chain".getBytes(StardardCharsets.US_ASCII)`

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11540/commits/e3f3413dae7f676d326d03e9a351d70cf2777f57

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/tokenmanager/AccessTokenManager.java:
```diff
@@ -0,0 +1,61 @@
+
+/** Manages access tokens for authenticating to the S2A. */
+@ThreadSafe
+public final class AccessTokenManager {
```
We should move all the subpackages that aren't supposed to be used into an internal subpackage, like `io.grpc.s2a.internal.handshaker.tokenmanager` here.

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11541

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/tokenmanager/AccessTokenManager.java:
```diff
@@ -0,0 +1,61 @@
+  private final TokenFetcher tokenFetcher;
+
+  /** Creates an {@code AccessTokenManager} based on the environment where the application runs. */
+  @SuppressWarnings("RethrowReflectiveOperationExceptionAsLinkageError")
```
Why was this ignored?

rmehta19 replied:
Un-ignoring in https://github.com/grpc/grpc-java/pull/11540/commits/81ea900590c13ce38e2212c089f9428444b89b35

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/tokenmanager/AccessTokenManager.java:
```diff
@@ -0,0 +1,61 @@
+    Optional<?> tokenFetcher;
+    try {
+      Class<?> singleTokenFetcherClass =
+          Class.forName("io.grpc.s2a.handshaker.tokenmanager.SingleTokenFetcher");
```
Why is this loaded with reflection?

rmehta19 replied:
We may add different types of token fetchers in the future (if we do S2A integration in a new environment), and if we can't load one, we can try to load another one.

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AStub.java:
```diff
@@ -0,0 +1,221 @@
+    @Override
+    public void onNext(SessionResp resp) {
+      verify(!doneReading);
+      responses.offer(Result.createWithResponse(resp));
```
All the `offer()`s scare me. The queue is sort of randomly sized (not just 2), which also doesn't feel me with confidence. It'd be much better to use `add()` which throws an exception if the queue is full instead of silently losing data.

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11540/commits/d13bd88b66d4bb24d1879b3afe1e961bf212fb96

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AStub.java:
```diff
@@ -0,0 +1,221 @@
+        exception = e;
+      }
+      responses.clear();
+      if (exception != null) {
```
Why do you allow non-exception results here? If there was ever a non-exception present, that _guarantees_ 1) that this code doesn't know what is going on and 2) there is a race (as that message could have been delayed until send() sent the request, at which point it looks like it is the response for the wrong request).

It seems clear to me you've mirrored some of the ALTS code, but I don't know why it was made complicate/worse in ways like this.

rmehta19 replied:
This logic existed before we added ```withWaitForReady``` on the S2AStub. We had this logic to handle a very specific case:
sometimes the client tries sending the first RPC before the S2A is healthy, so the response returned was always the response to the first RPC. This hasn't caused any problems yet. But now that we are using ```withWaitForReady```, we don't need this type of logic, so I've removed it in: https://github.com/grpc/grpc-java/pull/11540/commits/9efaa6a1fa1c3de02267f373950698a6621562d1

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/ProtoUtil.java:
```diff
@@ -0,0 +1,96 @@
+   * @return a {@link String} representing the ciphersuite.
+   * @throws AssertionError if the {@link Ciphersuite} is not one of the supported ciphersuites.
+   */
+  static String convertCiphersuite(Ciphersuite ciphersuite) {
```
This looks like dead code.

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11540/commits/1e8485cd687738000169efea93e4c248fa4e5259

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/ProtoUtil.java:
```diff
@@ -0,0 +1,96 @@
+   * @return a {@link String} representation of the TLS version.
+   * @throws AssertionError if the {@code tlsVersion} is not one of the supported TLS versions.
+   */
+  static String convertTlsProtocolVersion(TLSVersion tlsVersion) {
```
`@VisibleForTesting`? Looks like it is only used privately.

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/ProtoUtil.java:
```diff
@@ -0,0 +1,96 @@
+      case TLS_VERSION_1_0:
+        return "TLSv1";
+      default:
+        throw new AssertionError(
```
Using an error for an unexpected argument is asking for huge trouble. From what I can tell, this is not currently checked by the caller _and_ this comes from a remote system.

rmehta19 replied:
Yes, the tls version info is coming from S2A. Would it be appropriate if we threw an ```IllegalArgumentException``` here, and dealt with it in the caller (SslContextFactory)?

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11588/commits/9fdaa3684369dd6112d887e3dea2893d4573f61f

ejona86 reviewed (commented):
More random stuff. This is mostly from just skimming randomly and noticing unexpected code shapes.

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java:
```diff
@@ -0,0 +1,249 @@
+
+      Channel ch = channelPool.getChannel();
+      S2AServiceGrpc.S2AServiceStub stub = S2AServiceGrpc.newStub(ch);
+      S2AStub s2aStub = S2AStub.newInstance(stub);
```
This Stub is never closed. This causes IntegrationTest (after Larry's reordering of awaitTermination) to not complete tests quickly, as the RPC is outstanding so awaitTermination() hangs. I think once we shut down the top-level channel that releases the S2AHandshakerServiceChannel.ChannelResource after 1 second, which then calls shutdownNow() on the s2a channel which cancels the RPC.

rmehta19 replied:
Yes, we never call close on the stub directly. When the 
 [close](https://github.com/grpc/grpc-java/blob/master/s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java#L144) gets invoked, everything you mentioned happens.

ejona86 replied:
That's not okay. That means we leak an RPC per connection for the life of _all_ s2a channels, which is probably the lifetime of the process.

rmehta19 replied:
Would closing the stub in [close](https://github.com/grpc/grpc-java/blob/master/s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java#L146) resolve the issue?

rmehta19 replied:
The channel is still needed after the SslContext future is complete (we make RPCs to S2A to verify the peer, and perform private key operations during the handshake). The `SslContextBuilder` [configures the trustManager and PrivateKeyMethod with the stub](https://github.com/grpc/grpc-java/blob/master/s2a/src/main/java/io/grpc/s2a/internal/handshaker/SslContextFactory.java#L85-L88). In this example, the channel gets returned to the pool before the handshake is complete, which causes an error when S2ATrustManager uses the stub to verify the peer cert.

The stub is no longer needed when the handshake is complete. A probably not great option would be to pass around the channel instead, and create a new stub for every RPC? Or perhaps we plumb the stub down to the ClientTlsProtocolNegotiator and close it when the [handshake is done](https://github.com/grpc/grpc-java/blob/927d21541dbc34be245f09c8180358a94ae5c031/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L644)?

rmehta19 replied:
@ejona86, WDYT?

ejona86 replied:
I see. A new stub for every RPC is fine. I thought y'all _needed_ to use the same streaming RPC to preserve context. If separate RPCs are fine, that is definitely superior.

We don't want to inject into the handshake completion code itself, but that code is triggering a ProtocolNegotiationEvent. We can use that.

```java
private static final class S2AStubCleanupNegotiationHandler extends ProtocolNegotiationHandler {
  private final s2aStub; // pass in constructor
  @Override
  protected void protocolNegotiationEventTriggered(ChannelHandlerContext ctx) {
    s2aStub.close();
    fireProtocolNegotiationEvent(ctx);
    ctx.pipeline().remove(ctx);
  }
}
```

And then in S2aProtocolNegotiationHandler, `ctx.pipeline().addAfter()` that `S2AStubCleanupNegotiationHandler` before adding the tls handler. (So the TLS handler will be between that cleanup handler and the s2a handler.)

We could do alternatives like adding the tls handler _before_ the s2a handler, but they take a bit more reorganization.

rmehta19 replied:
Thanks Eric! I've implemented what you mention in https://github.com/grpc/grpc-java/pull/11600

I initially did think that we could do a new stub for each RPC, since we attach the localIdentity to each RPC, but chatted with @matthewstevenson88 and he mentioned there is a requirement that the lifetime of the bidirectional stream from the client to the S2A must be tied to the lifetime of the handshake.

rmehta19 commented:
Thanks for the comments @ejona86; I'll address them in separate commits at https://github.com/grpc/grpc-java/pull/11534. Let me know if you prefer separate PRs for them, or any other way of organizing.

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AHandshakerServiceChannel.java:
```diff
@@ -0,0 +1,195 @@
+      try {
+        isDelegateTerminated =
+            delegate.awaitTermination(DELEGATE_TERMINATION_TIMEOUT.getSeconds(), SECONDS);
+      } catch (InterruptedException e) {
```
Done in https://github.com/grpc/grpc-java/pull/11534/commits/2a169bacb33572c00a11204102d15a511d7c3603 .

rmehta19 replied:
Done in https://github.com/grpc/grpc-java/pull/11534/commits/c1375bb3d491cee3cc3ce3f175a23915a40dc38b

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,130 @@
+   * @param s2aAddress the address of the S2A server used to secure the connection.
+   * @return a {@code S2AChannelCredentials.Builder} instance.
+   */
+  public static Builder createBuilder(String s2aAddress) {
```
S2AChannelCredentials is used to create gRPC ChannelCredentials that are configured to use S2A. s2aAddress is required because without it we would not be able to talk to S2A to get info needed to complete the handshake. We default to a plaintext connection between the client (application which creates ChannelCredentials using S2AChannelCredentials.createBuilder("s2a_addr")) and S2A because mTLS to S2A is not available in all environments. If the client wants and is able to use mTLS to S2A, they can set the s2aChannelCredentials. Please let me know if I missed something in addressing this comment.

An aside: probably will be getting rid of setS2AChannelCredentials as we will merge MtlsToS2AChannelCredentials and S2AChannelCredentials as you suggested in above comment.

cc: @matthewstevenson88

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,130 @@
+   * @param s2aAddress the address of the S2A server used to secure the connection.
+   * @return a {@code S2AChannelCredentials.Builder} instance.
+   */
+  public static Builder createBuilder(String s2aAddress) {
```
Done in https://github.com/grpc/grpc-java/pull/11534/commits/467ddf55cca2640f64ca5b075ba96f886251d723

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/MtlsToS2AChannelCredentials.java:
```diff
@@ -0,0 +1,96 @@
+      File certChainFile = new File(certChainPath);
+      File trustBundleFile = new File(trustBundlePath);
+
+      AdvancedTlsX509KeyManager keyManager = new AdvancedTlsX509KeyManager();
```
Done in https://github.com/grpc/grpc-java/pull/11534/commits/9c44754f4280df9e99c422b33ec65782ea5e25aa

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,130 @@
+  public static final class Builder {
+    private final String s2aAddress;
+    private ObjectPool<Channel> s2aChannelPool;
+    private Optional<ChannelCredentials> s2aChannelCredentials;
```
Done in https://github.com/grpc/grpc-java/pull/11534/commits/ef53441de68e1f6da03bfc8f6726702dfda56c03

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,130 @@
+ * Configures gRPC to use S2A for transport security when establishing a secure channel. Only for
+ * use on the client side of a gRPC connection.
+ */
+public final class S2AChannelCredentials {
```
Done in https://github.com/grpc/grpc-java/pull/11534/commits/2a169bacb33572c00a11204102d15a511d7c3603

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2ATrustManager.java:
```diff
@@ -0,0 +1,152 @@
+    SessionResp resp;
+    try {
+      resp = stub.send(reqBuilder.build());
+    } catch (IOException | InterruptedException e) {
```
Done in https://github.com/grpc/grpc-java/pull/11534/commits/c1375bb3d491cee3cc3ce3f175a23915a40dc38b

ejona86 commented:
There's plenty of smaller/isolated items. Feel free to group those together as you wish, as they are easy to review. A few of the comments might turn into larger/plumbing changes. Some of those may be best to be separate. But overall, if you do lots of the "easy" stuff together that'd be fine to review/merge while you work on something more difficult.

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java:
```diff
@@ -0,0 +1,249 @@
+            @Override
+            public void onSuccess(SslContext sslContext) {
+              ChannelHandler handler =
+                  InternalProtocolNegotiators.tls(sslContext).newHandler(grpcHandler);
```
Thanks for pointing this out, although in order to do this, I think we need to add a new accessor to https://github.com/grpc/grpc-java/blob/9b0c19e698f0aa28b10f1c6513b5fb51b3e21954/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L685 from InternalProtocolNegotiators. 

Changes + adding the new accessor done in: https://github.com/grpc/grpc-java/pull/11539/commits/03ced7307168c3160c4117ba6ca43555a727ec0a

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AHandshakerServiceChannel.java:
```diff
@@ -0,0 +1,195 @@
+    @Override
+    public Channel create() {
+      EventLoopGroup eventLoopGroup =
+          new NioEventLoopGroup(1, new DefaultThreadFactory("S2A channel pool", true));
```
Removed the usage of own event loop in https://github.com/grpc/grpc-java/pull/11539/commits/03ced7307168c3160c4117ba6ca43555a727ec0a

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AHandshakerServiceChannel.java:
```diff
@@ -0,0 +1,195 @@
+ */
+@ThreadSafe
+public final class S2AHandshakerServiceChannel {
+  private static final ConcurrentMap<String, Resource<Channel>> SHARED_RESOURCE_CHANNELS =
```
This was forked from alts, but we don't need it for S2A. I've removed the usage of the global in https://github.com/grpc/grpc-java/pull/11540/commits/bbfe93bb64ab97355f825ac60a9046de417d3833

rmehta19 commented:
@ejona86 , @larry-safran I have addressed all the comments on this PR in the following 3 PRs:

- https://github.com/grpc/grpc-java/pull/11534
- https://github.com/grpc/grpc-java/pull/11539
- https://github.com/grpc/grpc-java/pull/11540

PTAL at these 3 and leave any comments.

I will address the comment to [move things to internal package](https://github.com/grpc/grpc-java/pull/11113#discussion_r1762067678) and [combine the MTLS and non-MTLS apis](https://github.com/grpc/grpc-java/pull/11113#discussion_r1761896842) in the following 2 PRs, once the above 3 are merged.

- https://github.com/grpc/grpc-java/pull/11541
- https://github.com/grpc/grpc-java/pull/11544

Besides these changes, please let me know anything else we can do to enable s2a to be included in a grpc-java release. Thank you!

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AStub.java:
```diff
@@ -0,0 +1,221 @@
+        throw new IOException(
+            "Received an unexpected response from a host at the S2A's address. The S2A might be"
+                + " unavailable."
+                + exception.getMessage());
```
This should be the cause of the IOException, not just add the message. Adding the message loses the exception type, the stack trace, and any causes that `exception` has. `new IOException("Received an ... unavailable", exception)`.

rmehta19 replied:
Thanks for pointing this out, done in https://github.com/grpc/grpc-java/pull/11570

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/MtlsToS2AChannelCredentials.java:
```diff
@@ -0,0 +1,96 @@
+ * Configures an {@code S2AChannelCredentials.Builder} instance with credentials used to establish a
+ * connection with the S2A to support talking to the S2A over mTLS.
+ */
+public final class MtlsToS2AChannelCredentials {
```
Done in https://github.com/grpc/grpc-java/pull/11544

ejona86 commented on s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java:
```diff
@@ -0,0 +1,130 @@
+          SharedResourcePool.forResource(
+              S2AHandshakerServiceChannel.getChannelResource(s2aAddress, s2aChannelCredentials));
+      checkNotNull(s2aChannelPool, "s2aChannelPool");
+      this.s2aChannelPool = s2aChannelPool;
```
Try to avoid mutating builders in the `build()` method. There seems to be no need to use a field here. You can just pass this pool to `buildProtocolNegotiatorFactory()` or (better) have `buildProtocolNegotiatorFactory()` construct the `s2aChannelPool` itself. Or you inline `buildProtocolNegotiatorFactory()` into the `build()` method.

IIRC, ALTS had `buildProtocolNegotiatorFactory()` for use in the older GoogleDefaultChannelBuilder API and related. In your case, you don't need that, so you probably don't need the method split out.

rmehta19 replied:
Thanks for pointing that out, https://github.com/grpc/grpc-java/pull/11573 constructs the ```s2aChannelPool``` in ```buildProtocolNegotiatorFactory```

ejona86 commented on s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java:
```diff
@@ -0,0 +1,249 @@
+  @VisibleForTesting
+  static final class S2AProtocolNegotiator implements ProtocolNegotiator {
+
+    private final S2AChannelPool channelPool;
```
To avoid creating+recreating the Channel repeatedly, we can rely on the 1 second timer of SharedResourceHolder. But we have retry backoffs that start at 1 second, so that's probably too small. An approach we've done elsewhere is to have this ProtocolNegotiator hold an unused reference to the Channel. So in the constructor it grabs a reference and in `close()` it releases the reference. This keeps the Channel alive for the lifetime of this negotiator.

Since we don't actually want a connection that entire time, we then configure ManagedChannelBuilder.idleTimeout() with the amount of time it should wait before shutting down its connections and going to sleep.

On client-side, the ProtocolNegotiator lives for the life of all connections. So we could avoid each connection getting its own reference to the Channel. But on server-side connections can outlive the negotiator. So it still makes sense for the Handler to acquire its own reference from the pool. (Yes, there's no server-side now, but it seems we'd expect that to happen eventually.)

rmehta19 replied:
Thank you for the explanation, done in https://github.com/grpc/grpc-java/pull/11584/commits/f7574613753dea7ae4c178cc33336a858c0571c0.

rmehta19 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AGrpcChannelPool.java:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  @Override
+  public synchronized void close() {
```
Deleted the Class in https://github.com/grpc/grpc-java/pull/11584/commits/f7574613753dea7ae4c178cc33336a858c0571c0. The other thing this class provides is an upper limit on the allowed users of the channel(`MAX_NUMBER_USERS_OF_CACHED_CHANNEL`); does gRPC provide an upper limit on the number of users per channel, so that it’s ok we no longer have this check?

rmehta19 replied:
Ok, I'll add that the s2a server grpc layer

ejona86 commented on s2a/src/main/java/io/grpc/s2a/channel/S2AHandshakerServiceChannel.java:
```diff
@@ -0,0 +1,195 @@
+ */
+@ThreadSafe
+public final class S2AHandshakerServiceChannel {
+  private static final ConcurrentMap<String, Resource<Channel>> SHARED_RESOURCE_CHANNELS =
```
This map was removed in #11540
